File: solution/array/134-gas-station/solution.go
package main

func canCompleteCircuit(gas []int, cost []int) int {
	if len(gas) == 0 || len(cost) == 0 {
		return -1
	}

	totalBalance := 0
	currentTank := 0
	startStation := 0

	for i := 0; i < len(gas); i++ {
		fuelDifference := gas[i] - cost[i]

		totalBalance += fuelDifference
		currentTank += fuelDifference

		if currentTank < 0 {
			startStation = i + 1
			currentTank = 0
		}
	}

	if totalBalance < 0 {
		return -1
	}

	return startStation
}

File: solution/array/134-gas-station/solution.js
/**
 * @param {number[]} gas
 * @param {number[]} cost
 * @return {number}
 */
function canCompleteCircuit(gas, cost) {
    if (!gas || !cost || gas.length === 0) {
        return -1;
    }

    let totalBalance = 0;
    let currentTank = 0;
    let startStation = 0;

    for (let i = 0; i < gas.length; i++) {
        const fuelDifference = gas[i] - cost[i];

        totalBalance += fuelDifference;
        currentTank += fuelDifference;

        if (currentTank < 0) {
            startStation = i + 1;
            currentTank = 0;
        }
    }

    return totalBalance >= 0 ? startStation : -1;
}